README.md
# Unirest for Node.js [![Build Status][travis-image]][travis-url] 

[![License][npm-license]][license-url]
[![Downloads][npm-downloads]][npm-url]
[![Gitter][gitter-image]][gitter-url]

![][unirest-logo]


[Unirest](http://unirest.io) is a set of lightweight HTTP libraries available in multiple languages, built and maintained by [Mashape](https://github.com/Mashape), who also maintain the open-source API Gateway [Kong](https://github.com/Mashape/kong). 


## Installing

To utilize unirest for node.js install the the `npm` module:

```bash
$ npm install unirest
```

After installing the `npm` package you can now start simplifying requests like so:

```js
var unirest = require('unirest');
```

## Creating Requests

You're probably wondering how by using **Unirest** makes creating requests easier. Besides automatically supporting gzip, and parsing responses, lets start with a basic working example:

```js
unirest.post('http://mockbin.com/request')
.headers({'Accept': 'application/json', 'Content-Type': 'application/json'})
.send({ "parameter": 23, "foo": "bar" })
.end(function (response) {
  console.log(response.body);
});
```

## Uploading Files

Transferring file data has been simplified:

```js
unirest.post('http://mockbin.com/request')
.headers({'Content-Type': 'multipart/form-data'})
.field('parameter', 'value') // Form field
.attach('file', '/tmp/file') // Attachment
.end(function (response) {
  console.log(response.body);
});
```

## Custom Entity Body

```js
unirest.post('http://mockbin.com/request')
.headers({'Accept': 'application/json'})
.send(new Buffer([1,2,3]))
.end(function (response) {
  console.log(response.body);
});
```

# Unirest

A request can be initiated by invoking the appropriate method on the unirest object, then calling `.end()` to send the request. Alternatively you can send the request directly by providing a callback along with the url.

## unirest(method [, uri, headers, body, callback])

- `method` - Request type (GET, PUT, POST, etc...)
- `uri` - _Optional_; When passed will return a [Request](#request) object. Otherwise returns generated function with `method` pre-defined (e.g. `unirest.get`)
- `headers` (`Object`) - _Optional_; HTTP Request headers
- `body` (`Mixed`) - _Optional_; HTTP Request body
- `callback` (`Function`) - _Optional_; Invoked when Request has finalized with the argument [Response](#response)

## unirest\[method](url [, headers, body, callback])

- `method` - Request type, pre-defined methods, see below.
- `url` - Request location.
- `headers` (`Object` | `Function`) - _Optional_; When `Object` headers are passed along to the [`Request.header`](#requestheaderobject-or-field-value) method,
   when `Function` this argument is used as the `callback`.
- `body` (`Mixed` | `Function`) - _Optional_; When `body` is not a `Function` it will be passed along to `Request.send()` method,
   otherwise when a `Function` it will be used as the `callback`.
- `callback` (`Function`) - _Optional_; Calls end with given argument, otherwise `Request` is returned.

All arguments above, with the exclusion of `url`, will accept a `Function` as the `callback`.
When no `callback` is present, the [Request](#request) object will be returned.

### get

Returns a [Request](#request) object with the `method` option set to `GET`

```js
var Request = unirest.get('http://mockbin.com/request');
```

### head
Returns a [Request](#request) object with the `method` option set to `HEAD`

```js
var Request = unirest.head('http://mockbin.com/request');
```

### put
Returns a [Request](#request) object with the `method` option set to `PUT`

```js
var Request = unirest.put('http://mockbin.com/request');
```

### post
Returns a [Request](#request) object with the `method` option set to `POST`

```js
var Request = unirest.post('http://mockbin.com/request');
```

### patch

Returns a [Request](#request) object with the `method` option set to `PATCH`

```js
var Request = unirest.patch('http://mockbin.com/request');
```

### delete
Returns a [Request](#request) object with the `method` option set to `DELETE`

```js
var Request = unirest.delete('http://mockbin.com/request');
```

## unirest.jar()

Creates a container to store multiple cookies, i.e. a cookie jar.

```js
var CookieJar = unirest.jar();
CookieJar.add('key=value', '/'); // Cookie string, pathname / url
unirest.get('http://mockbin.com/request').jar(CookieJar);
```

## unirest.cookie(String)

Creates a cookie, see above for example.

## unirest.request

`mikeal/request` library (the underlying layer of unirest) for direct use.

# Request

Provides simple and easy to use methods for manipulating the request prior to being sent. This object is created when a
Unirest Method is invoked. This object contains methods that are chainable like other libraries such as jQuery and popular
request module Superagent (which this library is modeled after slightly).

**Example**

```js
var Request = unirest.post('http://mockbin.com/request');

Request.header('Accept', 'application/json').end(function (response) {
  ...
});
```

## Request Methods

Request Methods differ from Option Methods (See Below) in that these methods transform, or handle the data in a sugared way, where as Option Methods require a more _hands on_ approach.

#### Request.auth(Object) or (user, pass, sendImmediately)

Accepts either an `Object` containing `user`, `pass`, and optionally `sendImmediately`.

- `user` (`String`) - Authentication Username
- `pass` (`String`) - Authentication Password
- `sendImmediately` (`String`) - _Optional_; Defaults to `true`; Flag to determine whether Request should send the basic authentication header along with the request. Upon being _false_, Request will retry with a _proper_ authentication header after receiving a `401` response from the server (which must contain a `WWW-Authenticate` header indicating the required authentication method)

**Object**

```js
Request.auth({
  user: 'Nijiko',
  pass: 'insecure',
  sendImmediately: true
});
```

**Arguments**

```js
Request.auth('Nijiko', 'insecure', true);
```

#### Request.header(header[, value])

**Suggested Method for setting Headers**

Accepts either an `Object` containing `header-name: value` entries,
or `field` and `value` arguments. Each entry is then stored in a two locations, one in the case-sensitive `Request.options.headers` and the other on a private `_headers` object that is case-insensitive for internal header lookup.

- `field` (`String`) - Header name, such as `Accepts`
- `value` (`String`) - Header value, such as `application/json`

**Object**

```js
Request.headers({
  'Accept': 'application/json',
  'User-Agent': 'Unirest Node.js'
})
```

Note the usage of [`Request.headers`](#requestheaders) which is simply an alias to the `Request.header` method, you can also use [`Request.set`](#requestset) to set headers.

**Arguments**

```js
Request.header('Accept', 'application/json');
```

#### Request.part(Object)

**Experimental**

Similiar to `Request.multipart()` except it only allows one object to be passed at a time and does the pre-processing on necessary `body` values for you.

Each object is then appended to the `Request.options.multipart` array.

```js
Request.part({
  'content-type': 'application/json',
  body: { foo: 'bar' }
}).part({
  'content-type': 'text/html',
  body: '<strong>Hello World!</strong>'
});
```

#### Request.query(Object) or (String)

Serializes argument passed to a querystring representation.

Should `url` already contain a querystring, the representation will be appended to the `url`.

```js
unirest.post('http://mockbin.com/request')
.query('name=nijiko')
.query({
  pet: 'spot'
})
.end(function (response) {
  console.log(response);
});
```

#### Request.send(Object | String)

Data marshalling for HTTP request body data

Determines whether data mime-type is `form` or `json`.
For irregular mime-types the `.type()` method is used to infer the `content-type` header.

When mime-type is `application/x-www-form-urlencoded` data is appended rather than overwritten.

**JSON**

```js
unirest.post('http://mockbin.com/request')
.type('json')
.send({
  foo: 'bar',
  hello: 3
})
.end(function (response) {
  console.log(response.body);
})
```

**FORM Encoded**

```js
// Body would be:
// name=nijiko&pet=turtle
unirest.post('http://mockbin.com/request')
.send('name=nijiko')
.send('pet=spot')
.end(function (response) {
  console.log(response.body);
});
```

**HTML / Other**

```js
unirest.post('http://mockbin.com/request')
.set('Content-Type', 'text/html')
.send('<strong>Hello World!</strong>')
.end(function (response) {
  console.log(response.body);
});
```

#### Request.type(String)

Sets the header `Content-Type` through either lookup for extensions (`xml`, `png`, `json`, etc...) using `mime` or using the full value such as `application/json`.

Uses [`Request.header`](#requestheaderobject-or-field-value) to set header value.

```js
Request.type('application/json') // Content-Type: application/json
Request.type('json') // Content-Type: application/json
Request.type('html') // Content-Type: text/html
…
```

## Request Form Methods

The following methods are sugar methods for attaching files, and form fields. Instead of handling files and processing them yourself Unirest can do that for you.

#### Request.attach(Object) or (name, path)

`Object` should consist of `name: 'path'` otherwise use `name` and `path`.

- `name` (`String`) - File field name
- `path` (`String` | `Object`) - File value, A `String` will be parsed based on its value. If `path` contains `http` or `https` Request will handle it as a `remote file`.  If `path` does not contain `http` or `https` then unirest will assume that it is the path to a local file and attempt to find it using `path.resolve`. An `Object` is directly set, so you can do pre-processing if you want without worrying about the string value.

**Object**

```js
unirest.post('http://mockbin.com/request')
.header('Accept', 'application/json')
.field({
  'parameter': 'value'
})
.attach({
  'file': 'dog.png',
  'relative file': fs.createReadStream(path.join(__dirname, 'dog.png')),
  'remote file': unirest.request('http://google.com/doodle.png')
})
.end(function (response) {
  console.log(response.body);
})
```

**Arguments**

```js
unirest.post('http://mockbin.com/request')
.header('Accept', 'application/json')
.field('parameter', 'value') // Form field
.attach('file', 'dog.png') // Attachment
.attach('remote file', fs.createReadStream(path.join(__dirname, 'dog.png')))  // Same as above.
.attach('remote file', unirest.request('http://google.com/doodle.png'))
.end(function (response) {
  console.log(response.body);
});
```

#### Request.field(Object) or (name, value)

`Object` should consist of `name: 'value'` otherwise use `name` and `value`

See `Request.attach` for usage.

#### Request.stream()

Sets `_stream` flag to use `request` streaming instead of direct `form-data` usage.
This seemingly appears to only work for node servers, use streaming only if you are a hundred percent sure it will work.
Tread carefully.

## Request.options

The _options_ `object` is where almost all of the request settings live. Each option method sugars to a field on this object to allow for chaining and ease of use. If
you have trouble with an option method and wish to directly access the _options_ object
you are free to do so.

This object is modeled after the `request` libraries options that are passed along through its constructor.

* `url` (`String` | `Object`) - Url, or object parsed from `url.parse()`
* `qs` (`Object`) - Object consisting of `querystring` values to append to `url` upon request.
* `method` (`String`) - Default `GET`; HTTP Method.
* `headers` (`Object`) - Default `{}`; HTTP Headers.
* `body` (`String` | `Object`) - Entity body for certain requests.
* `form` (`Object`) - Form data.
* `auth` (`Object`) - See `Request.auth()` below.
* `multipart` (`Object`) - _Experimental_; See documentation below.
* `followRedirect` (`Boolean`) - Default `true`; Follow HTTP `3xx` responses as redirects.
* `followAllRedirects` (`Boolean`) - Default `false`; Follow **Non**-GET HTTP `3xx` responses as redirects.
* `maxRedirects` (`Number`) - Default `10`; Maximum number of redirects before aborting.
* `encoding` (`String`) - Encoding to be used on `setEncoding` of response data.
* `timeout` (`Number`) - Number of milliseconds to wait before aborting.
* `proxy` (`String`) - See `Request.proxy()` below.
* `oauth` (`Object`) - See `Request.oauth()` below.
* `hawk` (`Object`) - See `Request.hawk()` below
* `strictSSL` (`Boolean`) - Default `true`; See `Request.strictSSL()` below.
* `secureProtocol` (`String`) - See `Request.secureProtocol()` below.
* `jar` (`Boolean` | `Jar`) - See `Request.jar()` below.
* `aws` (`Object`) - See `Request.aws()` below.
* `httpSignature` (`Object`) - See `Request.httpSignature()` Below.
* `localAddress` (`String`) - See `Request.localAddress()` Below.
* `pool` (`Object`) - See `Request.pool()` Below.
* `forever` (`Boolean`) - Default `undefined`; See `Request.forever()` Below

## Request Option Methods

#### Request.url(String)

Sets `url` location of the current request on `Request.options` to the given `String`

```js
Request.url('http://mockbin.com/request');
```

#### Request.method(String)

Sets `method` value on `Request.options` to the given value.

```js
Request.method('HEAD');
```

#### Request.form(Object)

Sets `form` object on `Request.options` to the given object.

When used `body` is set to the object passed as a `querystring` representation and the `Content-Type` header to `application/x-www-form-urlencoded; charset=utf-8`

```js
Request.form({
  key: 'value'
});
```

#### Request.multipart(Array)

**Experimental**

Sets `multipart` array containing multipart-form objects on `Request.options` to be sent along with the Request.

Each objects property with the exclusion of `body` is treated as a header value. Each `body` value must be pre-processed if necessary when using this method.

```js
Request.multipart([{
  'content-type': 'application/json',
  body: JSON.stringify({
    foo: 'bar'
  })
}, {
  'content-type': 'text/html',
  body: '<strong>Hello World!</strong>'
}]);
```

#### Request.maxRedirects(Number)

Sets `maxRedirects`, the number of redirects the current Request will follow, on `Request.options` based on the given value.

```js
Request.maxRedirects(6)
```

#### Request.followRedirect(Boolean)

Sets `followRedirect` flag on `Request.options` for whether the current Request should follow HTTP redirects based on the given value.

```js
Request.followRedirect(true);
```

#### Request.timeout(Number)

Sets `timeout`, number of milliseconds Request should wait for a response before aborting, on `Request.options` based on the given value.

```js
Request.timeout(2000)
```

#### Request.encoding(String)

Sets `encoding`, encoding to be used on setEncoding of response data if set to null, the body is returned as a Buffer, on `Request.options` based on given value.

```js
Request.encoding('utf-8')
```

#### Request.strictSSL(Boolean)

Sets `strictSSL` flag to require that SSL certificates be valid on `Request.options` based on given value.

```js
Request.strictSSL(true);
```

#### Request.httpSignature(Object)

Sets `httpSignature`

#### Request.proxy(String)

Sets `proxy`, HTTP Proxy to be set on `Request.options` based on value.

```js
Request.proxy('http://localproxy.com');
```

#### Request.secureProtocol(String)

Sets the secure protocol to use:

```js
Request.secureProtocol('SSLv2_method');
// or
Request.secureProtocol('SSLv3_client_method');
```

See [openssl.org](https://www.openssl.org/docs/ssl/SSL_CTX_new.html) for all possible values.

#### Request.aws(Object)

Sets `aws`, AWS Signing Credentials, on `Request.options`

```js
Request.aws({
  key: 'AWS_S3_KEY',
  secret: 'AWS_S3_SECRET',
  bucket: 'BUCKET NAME'
});
```

#### Request.oauth(Object)

Sets `oauth`, list of oauth credentials, on `Request.options` based on given object.

```js
var Request = unirest.get('https://api.twitter.com/oauth/request_token');

Request.oauth({

CHANGELOG.md
# Changelog

### Next (Master - Unstable)

- fix `Content-Type` contains `json` check
- add support for request timing `Unirest.get(...).time().end()`
- add support for sending `json` arrays
- add `Unirest.Response.statusCodes` dictionary for response code sugar properties
- add check for [MSIE `1223` status code](http://stackoverflow.com/questions/10046972/msie-returns-status-code-of-1223-for-ajax-request)
- update `request` to `2.65.0`
- update documentation and typos
- remove `Unirest.params`
- lint code to adhere to `standard`

lib/request.js
// Third Party
var StringDecoder = require("string_decoder").StringDecoder
var request = require("request")
var FormData = require("form-data")
var Stream = require("stream")
var path = require("path")
var zlib = require("zlib")
var fs = require('fs')

// Interal
var Marshallers = require("./marshals")
var Constants = require('./constants')
var Helpers = require("./helpers")
var Unirest = require('./unirest')

/**
 * @module Request
 * @author Nijiko Yonskai
 */

/**
 * Request Class
 *
 * @class
 * @param {Object}   options  Request Options, supports all options in request
 */
function Request (options) {
  if (typeof options === "undefined") {
    throw new Error("Invalid usage, options is required.")
  }

  // Assign options to the instance
  this.options = options

  // Create lowercase header mapping
  if (this.options.headers) {
    this.lowercaseHeaders = Object
      .keys(this.options.headers)
      .map(function (key, index) {
        return key.toLowerCase()
      })
  }
}

/**
 * Basic Header Authentication Method
 *
 * Supports user being an Object to reflect Request
 * Supports user, password to reflect SuperAgent
 *
 * @param  {String|Object}  username
 * @param  {String}         password
 * @param  {Boolean}        immediate Send authentication details immediately
 *                                    Some services like Google require this option to be true.
 * @return {Object}
 */
Request.prototype.auth = function auth (username, password, immediate) {
  this.options.auth = Helpers.type(username) === "Object" ? username : {
    user: username,
    password: password,
    sendImmediately: immediate
  }

  return this
}

/**
 * Turn on multipart-form streaming
 *
 * @return {Object}
 */
Request.prototype.stream = function stream () {
  this.options.stream = true
  return this
}

/**
 * Add cookie to existing or internally created cookie jar.
 *
 * When no cookie jar exists, a cookie jar is created using
 * `Unirest.request.jar` with no storage options passed.
 *
 * @param  {String} name  Cookie name
 * @param  {String} value Cookie value
 * @param  {String} url   URL to be parsed for the domain
 * @return {Request}
 */
Request.prototype.cookie = function cookie (name, value, url) {
  if (!this.options.jar) {
    this.options.jar = Unirest.request.jar()
  }

  try {
    this.options.jar.setCookie(
      request.cookie(name + '=' + value),
      url
    )
  } catch (error) {
    throw new Error("Invalid cookie jar, please set cookies through your jar rather than unirest.")
  }

  return this
}

/**
 * Add querystring value to this requests url.
 *
 * Infers whether to append to existing querystring or
 * create querystring.
 *
 * API
 *
 * - `r.query -> value:Object -> r`: Marshals `value` to querystring representation
 *   before appending to url.
 * - `r.query -> value:String -> r`: Takes string value and appends to url.
 *
 * @param  {Object|String} value
 * @return {Request}
 */
Request.prototype.query = function query (value) {
  var ending

  if (Helpers.type(value) === "Object") {
    value = Marshallers.query.unmarshal(value)
  }

  if (!value || typeof value.length === "undefined") {
    return this
  }

  ending = this.options.url[this.options.url.length - 1]

  if (ending !== '&' || ending !== '?') {
    this.options.url += Helpers.contains(this.options.url, '?') ? '&' : '?'
  }

  this.options.url += value

  return this
}

/**
 * Add header to the request
 *
 * @param  {String} field Header field
 * @param  {String} value Header field value
 * @return {Request}
 */
Request.prototype.header =
Request.prototype.headers = function header (field, value) {
  var name
  var key

  if (Helpers.type(field) === "Object") {
    for (key in field) {
      if (field.hasOwnProperty(key)) {
        this.header(key, field[key])
      }
    }
  } else {
    name = this.hasHeader(field)

    if (!name) {
      this.lowercaseHeaders.push(field.toLowerCase())
      name = field
    }

    this.options.headers[name] = value
  }

  return this
}

/**
 * Sets the `X-Mashape-Key` header using {@link #header}
 *
 * @param  {String} value Mashape key
 * @return {Request}
 */
Request.prototype.mashape = function mashape (value) {
  return this.header('X-Mashape-Key', value)
}

/**
 * Tests whether some element in the array is the specified header name,
 * checks are case insensitive.
 *
 * @param  {String}  name Header name
 * @return {Boolean}
 */
Request.prototype.hasHeader = function has_header (name) {
  var query = name.toLowerCase()
  var index = this.lowercaseHeaders.length

  while(index--) {
    if (query === this.lowercaseHeaders[index]) {
      return this.lowercaseHeaders[index]
    }
  }

  return false
}

/**
 * Associates content type header with specified value. Should the specified
 * value not be a mimetype (e.g. does not contain a foreward slash: json, html, xml)
 * a mimetype lookup is performed.
 *
 * @param  {String} type
 * @return {Request}
 */
Request.prototype.type = function type (value) {
  this.header("Content-Type", Helpers.contains(value, '/') ?
    value :
    mime.lookup(value)
  )

  return this
}

/**
 * Attaches a field or file to the multipart-form request
 *
 * @param  {String} name
 * @param  {String} value
 * @param  {Object} options
 * @return {Request}
 */
Request.prototype.field = function field (name, value, options) {
  var serialized
  var length
  var index
  var key

  // Initialize form-data instance for later
  // submission instead of requests library
  if (!this.options.formData) {
    this.options.formData = new FormData()
  }

  options = options || {
    attachment: false
  }

  // r.field -> name:Object -> r
  if (Helpers.type(name) === "Object") {
    for (key in name) {
      if (name.hasOwnProperty(key)) {
        this.field(key, name[key], options)
      }
    }

    return this
  }

  // r.field -> name:String -> value:Array -> ?options:Object -> r
  if (Helpers.type(value) === "Array") {
    options.marshalled = true

    for (index = 0, length = value.length; index < length; i++) {
      if (serialized = Request._marshalFieldValue(value[i])) {
        this.field(name, serialized, options)
      }
    }

    return this
  }

  // r.field -> name:String -> value:Mixed -> ?options:Object -> r
  if (value != null) {
    if (options.attachment) {
      // Fetch only when contains http like pathway
      if (Helpers.contains(value, "http://") || Helpers.contains(value, "https://")) {
        value = request(item.value)
      // Only create streams when non-buffer
      } else if (!(value instanceof fs.FileReadStream || value instanceof Buffer)) {
        value = fs.createReadStream(path.resolve(value))
      }
    } else if (!options.marshalled) {
      value = Request._marshalFieldValue(value)
    }

    this.options.formData.append(name, value, options)
  }

  return this
}

/**
 * Takes multipart options and places them on `options.multipart` array.
 * Transforms body when an `Object` or _content-type_ is present.
 *
 * Example:
 *
 *      Unirest
 *        .get('http://google.com')
 *        .part({
 *          'content-type': 'application/json',
 *          body: {
 *            phrase: 'Hello'
 *          }
 *        })
 *        .part({
 *           body: {
 *             phrase: 'World'
 *           }
 *        })
 *        .part('Attachment')
 *        .end(function (response) {});
 *
 * @param  {Object|String} options When an Object, headers should be placed directly on the object,
 *                                 not under a child property.
 * @return {Request}
 */
Request.prototype.part = function part (data) {
  var options = {}
  var type

  if (!this.options.multipart) {
    this.options.multipart = []
  }

  if (Helpers.type(data) === "Object" && (data.body && (data.type || data["content-type"]))) {
    type = Helpers.type(data.type || data["content-type"])

    if (type) {
      data.body = Marshallers.body.unmarshal(type, data.body)
    } else if (typeof data.body === "object") {
      data.type = "application/json"
      data.body = Marshallers.body.unmarshal("application/json", data.body)
    }
  } else {
    data = {
      body: data
    }
  }

  this.options.multipart.push(data)
  return this
}

/**
 * Attach a file for a multipart-form request
 *
 * @param  {String} name
 * @param  {String} value
 * @return {Request}
 */
Request.prototype.attach = function attach (name, value, options) {
  options = options || {}
  options.attachment = true
  return this.field(name, value, options)
}

/**
 * Data marshalling for HTTP request body data
 *
 * Determines whether data type is `form` or `json`.
 * For irregular mime-types the `.type()` method is used to infer the `content-type` header.
 *
 * When mime-type contains `application/x-www-form-urlencoded`, data is appended
 * to existing payload rather than overwritten.
 *
 * @param  {Mixed} data
 * @return {Request}
 */
Request.prototype.send =
Request.prototype.body =
Request.prototype.payload = function send (data) {
  var type = this.options.headers[this.hasHeader("content-type")]
  var isJSON = type ? type.indexOf("json") === -1 : false
  var dataType = Helpers.type(data)
  var key

  if (dataType === "Object" && !Buffer.isBuffer(data)) {
    if (!type || !isJSON) {
      this.type("form")
      type = this.options.headers[this.hasHeader("content-type")]
      this.options.body = Marshallers.body.unmarshal(type, data)
    } else if (isJSON) {
      this.options.json = true

      // Merge objects should body already exists and is an Object
      if (this.options.body && Helpers.type(this.options.body) === "Object") {
        for (key in data) {
          if (data.hasOwnProperty(key)) {
            this.options.body[key] = data[key]
          }
        }
      } else {
        this.options.body = data
      }
    }
  } else if (dataType === "String") {
    if (!type) {
      this.type("form")
      type = this.options.headers[this.hasHeader("content-type")]
    }

    if (type.indexOf("application/x-www-form-urlencoded") !== -1) {
      this.options.body = this.options.body
        ? this.options.body + '&' + data
        : data
    } else {
      this.options.body = (this.options.body || '') + data
    }
  } else {
    this.options.body = data
  }

  return this
}

/**
 * Invokes the request library to start the request. Any callback passed
 * is invoked after the request has finalized.
 *
 * @param  {Function} callback
 * @return {Object}
 */
Request.prototype.end = function (callback) {
  this.callback = callback
  this.request = request(this.options, this._onRequestResponse.bind(this))
  this.request.on("response", this._onDataResponse.bind(this))

  return this.request
}

/**
 * Promise alias, invokes {@link #end}
 *
 * @param {Function} callback Success Handler
 * @return {Promise}
 */
Request.prototype.then = function () {
  var promise = new Promise(this.end.bind(this))
  return promise.then.apply(promise, arguments)
}

/**
 * Handles final response from request library
 *
 * @param  {Object} error    Request Error
 * @param  {Object} response Request response object
 * @param  {Mixed} body      Request payload
 */
Request.prototype._onRequestResponse = function (error, response, body) {
  if (error && !response) {
    return this._createError(error)
  }

  // Handle empty request object edge-case
  if (!response) {
    console.trace('Request returned empty response object without error. Please report to: http://github.com/mashape/unirest-nodejs')
    return callback(new Error("Request returned empty response object"))
  }

  return this._createResponse(response, body)
}

/**
 * Handle decoding response
 *
 * @param  {Object} response request object
 */
Request.prototype._onDataResponse = function (response) {
  var _on
  var unzip
  var stream
  var decoder

  if (/^(deflate|gzip)$/.test(response.headers['content-encoding'])) {
    unzip = zlib.createUnzip()
    stream = new Stream()
    _on = response.on

    // Ensure no error from missing request
    stream.req = response.req

    // Map unzip errors to stream errors
    unzip.on('error', function (error) {
      stream.emit('error', error)
    })

    // Start unzipping process
    response.pipe(unzip)

    // Ensure encoding is captured
    response.setEncoding = function (type) {
      decoder = new StringDecoder(type)
    }

    // Capture decompression and decode with captured encoding
    unzip.on('data', function (buffer) {
      var string

      if (!decoder) {
        return stream.emit('data', buffer)
      }

      if (string = decoder.write(buffer) && string.length) {
        stream.emit('data', string)
      }
    })

    // Map unzip ending
    unzip.on('end', function () {
      stream.emit('end')
    })

    // Replace response on emitter with a custom emitter
    response.on = function (type, next) {
      if ('data' == type || 'end' == type) {
        stream.on(type, next)
      } else if ('error' == type) {
        _on.call(response, type, next)
      } else {
        _on.call(response, type, next)
      }
    }
  }
}

/**
 * Handle request errors
 *
 * @param  {Object} error
 */
Request.prototype._createError = function (error) {
  this.callback(error)
}

/**
 * Handle building response object from request response
 *
 * @param  {Object} response Response object from request
 * @param  {Mixed} body      Response body
 */
Request.prototype._createResponse = function (response, body) {
  /**
   * The Response
   *
   * @type {Object}
   */
  var Response = response

  /**
   * The raw Response body.
   *
   * @member
   * @type {Object|String}
   */
  Response.rawBody = body || response.body

  /**
   * The Response headers collection.
   *
   * @member
   * @type {Headers}
   */
  Response.headers = response.headers

  /**
   * The parsed Response type
   *
   * @type {String}
   */
  Response.type = Helpers.type(Response.headers['content-type'], true) || "application/octet-stream"

  /**
   * The Response cookie collection.
   *
   * @member
   * @type {Collection}
   */
  Response.cookies = Marshallers.cookie.marshal(Response.headers['cookie'])

  /**
   * The Response body, marshalled by content type should a matching type be
   * found.
   *
   * @member
   * @type {Object}
   */
  Response.body = BodyMarshal.marshal(Response.type, body || response.body)

  /**
   * @member
   * @alias {@link .statuscode}
   * @type {Number}
   */
  Response.code = response.statusCode

  /**
   * @member
   * @alias {@link .statuscode}
   * @type {Number}
   */
  Response.status = response.statusCode

  /**
   * Status code range
   *
   * @member
   * @type {Number}
   */
  Response.statusRange = (Response.status / 100 | 0)

  /**
   * Flagescribing whether Response was within the INFO
   * status code range, (1xx).
   *
   * @member
   * @type {Boolean}
   */
  Response.info = (1 === Response.statusRange)

  /**
   * Flag describing whether Response was within the OK
   * status code range, (2xx).
   *
   * @member
   * @type {Boolean}
   */
  Response.ok = (2 === Response.statusRange)

  /**
   * Flag describing whether Response was within the Redirect
   * status code range, (3xx).
   *
   * @member
   * @type {Boolean}
   */
  Response.redirection = (3 === Response.statusRange)

  /**
   * Flag describing whether Response was within the Client Error
   * status code range, (4xx).
   *
   * @member
   * @type {Boolean}
   */
  Response.clientError = (4 === Response.statusRange)

  /**
   * Flag describing whether Response was within the Server Error
   * status code range, (4xx).
   *
   * @member
   * @type {Boolean}
   */
  Response.serverError = (5 === Response.statusRange)

  /**
   * Flag describing whether Response contains the Accepted
   * status code (202).
   *
   * @member
   * @type {Boolean}
   */
  Response.accepted = (202 === Response.status)

  /**
   * Flag describing whether Response contains the NO CONTENT
   * status code (204).
   *
   * @member
   * @type {Boolean}
   */
  Response.noContent = (204 === Response.status || 1223 === Response.status)

  /**
   * Flag describing whether Response contains the BAD REQUEST
   * status code (400).
   *
   * @member
   * @type {Boolean}
   */
  Response.badRequest = (400 === Response.status)

  /**
   * Flag describing whether Response has the UNAUTHORIZED
   * status code (401).
   *
   * @member
   * @type {Boolean}
   */
  Response.unauthorized = (401 === Response.status)

  /**
   * Flag describing whether Response has the NOT ACCEPTABLE
   * status code (406).
   *
   * @member
   * @type {Boolean}
   */
  Response.notAcceptable = (406 === Response.status)

  /**
   * Flag describing whether Response has the NOT FOUND
   * status code (404).
   *
   * @member
   * @type {Boolean}
   */
  Response.notFound = (404 === Response.status)

  /**
   * Flag describing whether Response has the FORBIDDEN
   * status code (403).
   *
   * @member
   * @type {Boolean}
   */
  Response.forbidden = (403 === Response.status)

  /**
   * Flag describing whether Response within an ERROR
   * status code range (4xx - 5xx).
   *
   * @member
   * @type {Boolean|StatusError}
   */
  Response.error = (4 === Response.statusRange || 5 === Response.statusRange) ?
    new StatusError(Response.status, Response.body) :
    false

  // Append Set-Cookies to list of Cookies
  Helpers.extend(Response.cookies, Marshallers.cookie.marshal(Response.headers["set-cookie"]))

  // Invoke callback
  this.callback(null, Response);
}

/**
 * Marshals multipart field values according to the type of value.
 *
 * Ignores `Streams`, `Buffers`, and `String` values. `Objects` are `JSON` stringified.
 * Everything else is converted to `String`.
 *
 * @param {Object} value
 */
Request._marshalFieldValue = function marshal_field_value (value) {
  if (value instanceof fs.FileReadStream || value instanceof Buffer || typeof value === "string") {
    return value
  }

  if (Helpers.type(value) === "Object") {
    return Marshallers.json.unmarshal(value)
  }

  return value.toString()
}

/**
 * Request library
 *
 * @type {Object}
 */
Request.request = request

/**
 * Add request option to Request method list
 *
 * @param {String} name      Option name
 * @param {String} reference Original method for aliases
 * @private
 */
function addRequestSugarMethod (name, reference) {
  Request.prototype[name] = reference ? Request.prototype[reference] : function (value) {
    this.options[name] = value
  }
}

/**
 * Implements additional promise(ify) methods that
 *
 * @param {String} name Promise method name
 * @private
 */
function addRequestPromiseMethod (name) {
  Request.prototype[name] = function () {
    var then = this.then()
    return then[name].apply(then, arguments)
  }
}

// Setup methods from request options
for (var index = 0; index < Constants.REQUEST_OPTIONS.length; index++) {
  var option = Constants.REQUEST_OPTIONS[index]
  var reference = null

  if (option.indexOf(':') !== -1) {
    option = option.split(':')
    reference = option[1]
    option = option[0]
  }

  addRequestSugarMethod(option, reference)
}

// Setup methods from promise options
for (var index = 0; index < Constants.PROMISE_METHODS.length; index++) {
  addRequestPromiseMethod(Constants.PROMISE_METHODS[index])
}

// Export
module.exports = Request